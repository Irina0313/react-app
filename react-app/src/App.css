body {
  background-color: black;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
}

#root {
  max-width: 1280px;
  width: 100%;
  margin: 0 auto;
  padding: 2rem;
  text-align: center;
  font-size: 10px;
  color: aliceblue;
}

.mainWrapper {

  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  row-gap: 20px;
}

.mainTitle {
  margin: 0;
  font-size: 5em;
  color: darkgoldenrod;
}

.loading {
  font-size: 3em;
  color: aliceblue;
  text-align: center;
  margin: 20px 0;
  animation: pulse 2s ease infinite;
}

@keyframes pulse {

  0%,
  100% {
    opacity: 0.1;
  }

  50% {
    opacity: 1;
  }

}