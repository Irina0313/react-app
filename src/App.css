.mainWrapper {
  margin: 0 auto;
  max-width: 1280px;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  row-gap: 40px;
}

.mainTitle {
  margin: 0;
  font-size: 5em;
  line-height: 1.1;
  color: darkgoldenrod;
}

.titleWrapper {
  max-width: 100%;
  display: flex;
  justify-content: space-between;
}

.errorBtn {
  width: 150px;
  font-size: 1.8em;
  align-self: flex-end;
}

.errorBtn:hover {
  background-color: darkgoldenrod;
  border-color: aliceblue;
  color: aliceblue;
}

.loading {
  font-size: 3em;
  color: aliceblue;
  text-align: center;
  margin: 20px 0;
  animation: pulse 2s ease infinite;
}

@keyframes pulse {

  0%,
  100% {
    opacity: 0.1;
  }

  50% {
    opacity: 1;
  }
}

@media screen and (max-width: 1300px) {
  .titleWrapper {
    flex-direction: column;
    row-gap: 10px;
  }
}

@media screen and (max-width: 500px) {
  .errorBtn {
    padding: 8px;
    width: 100px;
    font-size: 1.8em;
    align-self: flex-end;
  }
}