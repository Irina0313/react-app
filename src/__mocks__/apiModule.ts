import { productsApi, IApiResp, IGetProducts } from '../store/productsAPI';
import { ProductProps } from '../Components/Data/Product';

export const mockFetchData = async (
  args: IGetProducts
): Promise<IApiResp | undefined> => {
  try {
    const result = productsApi.endpoints.getProducts.useQuery(args).data;
    if (result) {
      return result;
    }
  } catch (error) {
    throw error;
  }
};

export const mockFetchProductById = async (
  productId: number
): Promise<ProductProps | undefined> => {
  try {
    const result = productsApi.endpoints.getProductById.useQuery({
      productId,
    }).data;

    if (result) {
      return result;
    }
  } catch (error) {
    throw error;
  }
};
