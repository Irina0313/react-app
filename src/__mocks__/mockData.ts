import { IApiResp } from '../store/productsAPI';

export const mockData: IApiResp = {
  products: [
    {
      id: 20,
      title: 'iPhone 9',
      description: 'An apple mobile which is nothing like apple',
      price: 549,
      discountPercentage: 2.92,
      rating: 4.92,
      stock: 54,
      brand: 'Golden',
      category: 'smartphones',
      thumbnail: 'https://i.dummyjson.com/data/products/30/thumbnail.jpg',

      images: ['https://i.dummyjson.com/data/products/1/1.jpg'],
    },
  ],
  total: 100,
  skip: 0,
  limit: 30,
};
