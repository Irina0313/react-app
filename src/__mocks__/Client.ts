import { IApiResp } from '../api/Client';
import { mockContext } from './mockContext';
import { isMockServerError, isMockServerNoDataResponse } from '../App.test';

type MockClientReturnType = IApiResp | IErrorResp | null;

interface IErrorResp {
  products: [];
}

const errorResp: IErrorResp = {
  products: [],
};

const mockClient: jest.Mock<Promise<MockClientReturnType>, []> = jest.fn(() => {
  if (isMockServerNoDataResponse) {
    return Promise.resolve(errorResp);
  }
  if (isMockServerError) {
    throw new Error('Mock Server Error');
  }

  return Promise.resolve(mockContext);
});
export default mockClient;
