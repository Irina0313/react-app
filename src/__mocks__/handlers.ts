import { http, HttpResponse } from 'msw';
import { mockData } from './mockData';
import { mockProduct } from './mockProduct';

/* const products = 'https://dummyjson.com/products/'; */

export const handlers = [
  http.get('https://dummyjson.com/products/', () => {
    console.log('https://dummyjson.com/products/');

    return HttpResponse.json(mockData);
  }),
  http.get('https://dummyjson.com/products/20', () => {
    console.log('https://dummyjson.com/products/20');
    return HttpResponse.json(mockProduct);
  }),
];
