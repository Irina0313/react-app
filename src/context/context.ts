import { createContext } from 'react';
import { IApiResp } from '../api/Client';

interface SearchContextType {
  onSearch: (searchQuery: string | null) => void;
  prevSearchParams: string | null;
}

export const ProductsContext = createContext<IApiResp | null>(null);
export const LoadingContext = createContext<boolean>(false);
export const SearchContext = createContext<SearchContextType>({
  onSearch: () => {},
  prevSearchParams: null,
});
