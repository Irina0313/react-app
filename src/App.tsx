import { useState, useCallback, useEffect } from 'react';
import { Routes, Route, useNavigate, useLocation } from 'react-router-dom';
import ProductPage from './pages/ProductPage/ProductPage';
import NotFoundPage from './pages/NotFoundPage/NotFoundPage';
import Layout from './Components/Layout/Layout';
import client from './api/Client';
import './Components/Layout/Layout.css';
import HomePage from './pages/HomePage/HomePage';
import { IApiResp } from './api/Client';
import { ProductsContext, LoadingContext, SearchContext } from './context';
import safeJsonParse from './utils/JsonActions';

function App() {
  const navigate = useNavigate();

  const { pathname } = useLocation();
  const [data, setData] = useState<IApiResp | null>(null);
  const [loading, setLoading] = useState(false);
  const [showError, setShowError] = useState(false);
  const [searchParams, setSearchParams] = useState<string | null>(() =>
    safeJsonParse(localStorage.savedSearch)
  );
  const [isDataLoaded, setIsDataLoaded] = useState(false);
  const [itemsPerPage, setItemsPerPage] = useState(30);

  const segments = pathname.split('/');
  const [currPageNum, setCurrPageNum] = useState(
    (segments[1].length > 5 &&
      (() => {
        return Number(segments[1].slice(5));
      })) ||
      null
  );

  const loadProducts = useCallback(
    async (
      searchQuery: string | null = searchParams,
      items: number = itemsPerPage,
      currPageNum: number | null
    ) => {
      if (currPageNum) {
        try {
          setLoading(true);

          const data = await client(
            'products',
            searchQuery,
            items,
            currPageNum
          );
          setLoading(false);

          data && setData(data);
        } catch (error) {
          console.error('catch ', error);
          setLoading(false);
          setShowError(true);
        }
      } else {
        setLoading(false);
        setShowError(true);
      }
    },
    [itemsPerPage, searchParams]
  );

  useEffect(() => {
    if (!isDataLoaded) {
      (async () => {
        setIsDataLoaded(true);
        if (segments[1] === '') {
          navigate(`page=1`);
          setCurrPageNum(1);
          await loadProducts(searchParams, itemsPerPage, 1);
        } else {
          navigate(`${pathname}`);
          await loadProducts(searchParams, itemsPerPage, currPageNum);
        }
      })();
    }
  }, [
    searchParams,
    loadProducts,
    isDataLoaded,
    itemsPerPage,
    navigate,
    pathname,
    segments,
    currPageNum,
  ]);

  const handleSearch = async (searchQuery: string | null) => {
    setShowError(false);
    setLoading(true);

    localStorage.setItem('savedSearch', JSON.stringify(searchQuery));
    setSearchParams(searchQuery);
    await loadProducts(searchQuery, itemsPerPage, 1);
  };

  const handleTestError = () => {
    setShowError(true);
  };

  const handlePaginatorBtnsClick = async (
    pageNumber: number,
    items: number = 30
  ) => {
    setLoading(true);

    setItemsPerPage(items);
    setCurrPageNum(pageNumber);
    await loadProducts(searchParams, items, pageNumber);
  };

  const updateProducts = async () => {
    await loadProducts(searchParams, itemsPerPage, currPageNum);
  };

  return (
    <LoadingContext.Provider value={loading}>
      <SearchContext.Provider
        value={{ onSearch: handleSearch, prevSearchParams: searchParams }}
      >
        <ProductsContext.Provider value={data}>
          <Routes>
            <Route
              path={`/`}
              element={<Layout handleTestError={handleTestError} />}
            >
              <Route
                path={':page'}
                element={
                  <HomePage
                    showError={showError}
                    handlePaginatorBtnsClick={handlePaginatorBtnsClick}
                  />
                }
              >
                <Route
                  path={':productId'}
                  element={<ProductPage getProducts={updateProducts} />}
                />
              </Route>
            </Route>
            <Route path="*" element={<NotFoundPage />} />
          </Routes>
        </ProductsContext.Provider>
      </SearchContext.Provider>
    </LoadingContext.Provider>
  );
}

export default App;
