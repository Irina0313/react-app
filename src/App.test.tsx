import {
  render,
  screen,
  act,
  waitFor,
  fireEvent,
} from '@testing-library/react';
import App from './App';
import { BrowserRouter } from 'react-router-dom';
import { SearchContext, LoadingContext, ProductsContext } from './context';
import { mockContext } from './__mocks__/mockContext';
import { mockSetItem } from './__mocks__/mockLocalStorage';
import '@testing-library/jest-dom';

export let isMockServerError = false;
export let isMockServerNoDataResponse = false;

jest.mock('./api/Client', () => require('./__mocks__/Client'));

describe('App tests', () => {
  beforeEach(() => {
    mockSetItem.mockClear();
    jest.clearAllMocks();
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  const loading = false;
  let mockSearchParams = 'mySearch';
  const mockUpdateProducts = jest.fn();

  function initializePage() {
    jest.clearAllMocks();
    render(
      <BrowserRouter>
        <LoadingContext.Provider value={loading}>
          <SearchContext.Provider
            value={{ onSearch: () => {}, prevSearchParams: mockSearchParams }}
          >
            <ProductsContext.Provider value={mockContext}>
              <App />
            </ProductsContext.Provider>
          </SearchContext.Provider>
        </LoadingContext.Provider>
      </BrowserRouter>
    );
  }

  test('it should display right page numder and items per page when users use pagination', async () => {
    isMockServerError = false;
    mockSearchParams = '';
    await act(async () => {
      initializePage();
    });
    await waitFor(() => {
      const nextBtn = screen.getByText(/next/i);
      const prevBtn = screen.getByText(/prev/i);
      const itemsPerPage = screen.getByRole('combobox');

      expect(itemsPerPage).toBeTruthy();
      fireEvent.select(itemsPerPage, { value: '10' });
      waitFor(() => {
        expect(screen.getByDisplayValue('10')).toBeTruthy();
      });
      expect(prevBtn).toBeDisabled();
      expect(prevBtn).toBeTruthy();
      fireEvent.click(prevBtn);
      expect(screen.getByTestId('currPageNumber')).toHaveTextContent('1');
      expect(nextBtn).toBeTruthy();
      fireEvent.click(nextBtn);
      expect(screen.getByTestId('currPageNumber')).toHaveTextContent('2');
      fireEvent.select(itemsPerPage, { value: '10' });
    });
  });
  test('Check that clicking triggers an additional API call to fetch detailed information.', async () => {
    await act(async () => {
      initializePage();
    });
    await waitFor(() => {
      const testPageLink = screen.getAllByRole('link')[0];
      expect(testPageLink).toBeTruthy();
      fireEvent.click(testPageLink);
      waitFor(() => {
        expect(mockUpdateProducts).toHaveBeenCalled();
      });
    });
  });

  test('Validate that clicking on a card opens a detailed card component;', async () => {
    await act(async () => {
      initializePage();
    });
    await waitFor(() => {
      const testPageLink = screen.getAllByRole('link')[0];
      expect(testPageLink).toBeTruthy();
      fireEvent.click(testPageLink);
      waitFor(() => {
        const detailedPage = screen.getByTestId('modalBackdrop');
        expect(detailedPage).toBeTruthy();
      });
    });
  });
  test('renders the app with BrowserRouter', async () => {
    await act(async () => {
      initializePage();
    });

    expect(screen.getByText(/RSS React App Catalog/i)).toBeTruthy();
  });

  test('App snapshot', async () => {
    const app = await act(async () => {
      initializePage();
    });

    expect(app).toMatchSnapshot();
  });

  test('it should render correctly when the server return no data', async () => {
    isMockServerNoDataResponse = true;

    await act(async () => {
      initializePage();
    });

    await act(async () => {
      expect(screen.getByText(/Page was not found/i)).toBeTruthy();
      isMockServerError = false;
    });
  });

  test('should render error message when the server request fails', async () => {
    isMockServerError = true;

    await act(async () => {
      initializePage();
    });

    expect(screen.queryByText(/Loading.../i)).toBeFalsy();
    expect(screen.getByText(/Page was not found/i)).toBeTruthy();

    isMockServerError = false;
  });

  test('Verify that clicking the Search button saves the entered value to the local storage and Check that the component retrieves the value from the local storage upon mounting.', async () => {
    await act(async () => {
      initializePage();
    });

    const searchInput = screen.getByRole('searchbox');
    fireEvent.change(searchInput, { target: { value: 'mock search text  ' } });
    fireEvent.click(screen.getByRole('button', { name: 'Search' }));

    /*save search to the locale storage*/
    await waitFor(() => {
      const displayedSearchValue = screen.getByDisplayValue('mock search text');
      expect(displayedSearchValue).toBeTruthy();
      expect(mockSetItem).toHaveBeenCalledTimes(1);
      expect(mockSetItem).toHaveBeenCalledWith(
        'savedSearch',
        JSON.stringify('mock search text')
      );
    });

    /*initiate page reload*/
    await act(async () => {
      initializePage();
    });

    /*check that the Search input value component retrieves the value from the local storage upon mounting*/
    const displayedSearchValue = screen.getByDisplayValue('mock search text');
    expect(displayedSearchValue).toBeTruthy();
  });

  test('should handle test error', async () => {
    await act(async () => {
      initializePage();
    });

    await act(async () => {
      const testErrorBtn = screen.getByRole('button', { name: 'Test error' });
      fireEvent.click(testErrorBtn);
    });

    await screen.findByText(/Page was not found/i);

    const errorDisplayed = screen.getByText(/Page was not found/i);
    expect(errorDisplayed).toBeTruthy();
  });
});
