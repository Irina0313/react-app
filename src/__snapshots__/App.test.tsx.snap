// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests App snapshot 1`] = `undefined`;
